### PR TITLE
[NOT READY FOR REVIEW]Batch processing replication task queue POC

### PR DESCRIPTION
**Pull request by xwduan**
**What changed?**
Implement an sequential queue that has a batching functionality.

**Why?**
Improve the efficiency when processing replication task. If replication task can be batched, we can reduce DB i/o when processing replication tasks.

**How did you test it?**
N/a.

**Potential risks**
N/a

**Is hotfix candidate?**
no.

### CODE REVIEW

File: service/history/replication/batchable_task.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	"time"
+
+	"go.temporal.io/server/common/backoff"
+	ctasks "go.temporal.io/server/common/tasks"
+)
+
+type (
+	BatchableTask interface {
+		TrackableExecutableTask
+		Combine(task BatchableTask) (BatchableTask, error)
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/batchable_task.go:15]: Take HistoryEventReplicationTask as an example: TaskA contains events [1,2,3] TaskB contains events[4,5,6], the TaskA.Combine(TaskB) should return a new TaskC that contains events [1,2,3,4,5,6].

File: service/history/replication/sequential_batch_queue.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+package replication
+
+import (
+	"sync"
+
+	"go.temporal.io/server/common/collection"
+	ctasks "go.temporal.io/server/common/tasks"
+)
+
+// SequentialBatchableTaskQueue is not thread-safe. It is a 2-tiered queue. If incoming task is batchable,
```
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/sequential_batch_queue.go:34]: Yeah, it has lock and technically it is thread safe, add and remove can happen in different threads. But in terms of usage, Remove and processing Task should be synchronize:
It only support **single** thread doing this:
```
While (!q.isEmpty()) {
    task = q.Remove()
    task.Execute()
}
```
If multiple threads do above loop at same time, the issue would be, threadA may get BatchedTask[1,2,3] while threadB get IndividualTask[1].

File: service/history/replication/batchable_task.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/replication/batchable_task.go:15]: can you also expose `func IsCombine() bool` API?

File: service/history/replication/batchable_task.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	}
+
+	// batchedTask should be only used by sequential_combined_queue
+	batchedTask struct {
+		batchedTask     BatchableTask
+		individualTasks []BatchableTask
```
Comment (wxing1292) [inline service/history/replication/batchable_task.go:21]: should this line be `individualTasks [] TrackableExecutableTask`?

File: service/history/replication/batchable_task.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+func (w *batchedTask) TaskCreationTime() time.Time {
+	return w.individualTasks[0].TaskCreationTime()
+}
+
+func (w *batchedTask) MarkPoisonPill() error {
+	if len(w.individualTasks) == 1 {
+		return w.batchedTask.MarkPoisonPill()
+	} else {
+		return nil
```
Comment (wxing1292) [inline service/history/replication/batchable_task.go:52]: this should not return nil, but rather attempting to break the batched task and apply one by one?
plz add a TODO here

Comment (wxing1292) [reply service/history/replication/batchable_task.go:52]: or `len(w.individualTasks) > 1` is not expected? then logic should panic here

File: service/history/replication/batchable_task.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+func (w *batchedTask) Cancel() {
+	w.callIndividual(BatchableTask.Cancel)
+}
+
+func (w *batchedTask) Nack(err error) {
+	if len(w.individualTasks) == 1 {
+		w.batchedTask.Nack(err)
+	} else {
+		// TODO: logging
```
Comment (wxing1292) [inline service/history/replication/batchable_task.go:92]: https://github.com/temporalio/temporal/pull/4804/files#r1308057946

File: service/history/replication/batchable_task.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+	}
+}
+
+func (w *batchedTask) Reschedule() {
+	if len(w.individualTasks) == 1 {
```
Comment (wxing1292) [inline service/history/replication/batchable_task.go:97]: what if `len(w.individualTasks) > 1`?

File: service/history/replication/batchable_task.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	}
+}
+
+func (w *batchedTask) addTask(task BatchableTask) error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	// This is to make sure no task can be added into this batch after it starts executing
+	if w.state != batchStateOpen {
+		return errors.New("cannot combine tasks on non-pending state")
```
Comment (wxing1292) [inline service/history/replication/batchable_task.go:118]: do not use errors.new

if a batched task is sealed, i guess you should return false here

File: service/history/replication/batchable_task.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	newTask, err := w.batchedTask.Combine(task)
+	if err != nil {
+		return err
+	}
```
Comment (wxing1292) [inline service/history/replication/batchable_task.go:123]: if w.batchedTask. CanCombine(task) {
    newTask, err := w.batchedTask.Combine(task)
} else {
    return false
}

File: service/history/replication/sequential_batch_queue.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+func (q *SequentialBatchableTaskQueue) ID() interface{} {
+	return q.id
+}
+
+func (q *SequentialBatchableTaskQueue) Peek() TrackableExecutableTask {
+	q.Lock()
+	defer q.Unlock()
+	if q.mainQueue.IsEmpty() {
+		return q.individualTaskQueue[0]
```
Comment (wxing1292) [inline service/history/replication/sequential_batch_queue.go:70]: panic?

**Pull request closed without merging**